Syscall Sysv: Add support for unreferencing callbacks

pdf opened the pull request:
Also deduplicate callbacks by function pointer.

Function equality in Go is problematic, which is why functions are not comparable in the language spec or via reflect.Value.Equal().

For our particular use-case though I believe that comparing function pointers is good enough, since we don't need to differentiate closures based on scope variable content.

jwijenbergh commented:
Wow! I think this is exactly what we want. The notion of slots/holes was my idea as well. It generally looks good, if these tests pass I would say LGTM

pdf commented:
Hmm, not sure what's up with the test runners, looks like the FreeBSD runners are in a boot loop. I would like to see the tests pass for platforms/architectures I don't have access to, but they pass locally on amd64 with 1.19 / 1.20 / 1.21.

jwijenbergh commented on syscall_sysv.go:
```diff
@@ -47,10 +88,22 @@ func NewCallback(fn interface{}) uintptr {
+	knownFn  map[uintptr]uintptr
+	knownIdx map[uintptr]int
+}{
+	holes:    make(map[int]struct{}),
```
nit: specify the length here (maxCB) as the second argument to make

jwijenbergh commented on syscall_sysv.go:
```diff
@@ -47,10 +88,22 @@ func NewCallback(fn interface{}) uintptr {
-	funcs [maxCB]reflect.Value // the saved callbacks
+var cbs = struct {
+	lock     sync.RWMutex
+	holes    map[int]struct{}
+	funcs    [maxCB]reflect.Value // the saved callbacks
+	knownFn  map[uintptr]uintptr
+	knownIdx map[uintptr]int
```
Could you add some comments what these do? knownFn and knownIdx are a bit vague on first glance

jwijenbergh commented:
> Hmm, not sure what's up with the test runners, looks like the FreeBSD runners are in a boot loop. I would like to see the tests pass for platforms/architectures I don't have access to, but they pass locally on amd64 with 1.19 / 1.20 / 1.21.

let me rebase with newest purego anyways. But I was mainly commenting that because the ubuntu 1.19.x container failed

edit: done, I should probably not force push the main branch in the future now that the use case is n > 1, sorry!

pdf commented:
Updated per review comments and rebased against main.

jwijenbergh commented:
Thanks! I can reproduce the CI crashes with:

```
CGO_ENABLED=1 go test -shuffle=on -v -count=10
```

any idea? I can have a more detailed look later

pdf commented:
This is related to callback re-use - I believe what's happening is that when the callback is re-used across goroutines it logs to the originally captured test instance, which results in logs being output to a comlpeted test run.

Either removing the log line inside the callback, or unref'ing the cb inside the callback and creating a new cb for each iteration allows the tests to pass, but this does illustrate a pitfall with callback re-use (and function equality in general) for closures.

pdf commented:
I tested a number of scenarios, but didn't find this one - I think perhaps this means that the concept of callback re-use is flawed, since it cannot work correctly in the presence of changing variables passed into the outer scope for the callback. The below should illustrate a little more clearly.

This will work:
```go
for i := 0; i < 10_000; i++ {
	time.Sleep(10 * time.Millisecond)

	cb := func(data uintptr) bool {
		fmt.Println(`i is increasing in scope`, i)
		return false
	}
	glib.IdleAdd(cb, 0)
}
```

But this will not work:
```go
for i := 0; i < 10_000; i++ {
	time.Sleep(10 * time.Millisecond)

	update(i)
}

func update(i int) {
	cb := func(data uintptr) bool {
		fmt.Println(`i is captured from scope at first definition and not increasing`, i)
		return false
	}

	glib.IdleAdd(cb, 0)
}
```

We could remove this optimization while retaining manual unref'ing, or providing a separate API (e.g. `purego.NewReusableCallback()`) to enable re-use at the user's discretion, however the implication for our desired use-case in puregotk https://github.com/jwijenbergh/puregotk/pull/2 is that we don't have a safe way to track wrapped callbacks in the case that they might be instantiated with this kind of scope.

EDIT: Actually, not being able to rely on taking a pointer to the function value as a key invalidates the ability to unref callbacks by function value.

I think we could still support all of this functionality, but we would have to require that users pass in pointers to functions for it to work as expected in all scenarios (though I need to actually test this, I'm pretty confident it would work). I suspect that means going back to duplicating all of the callback-accepting methods in puregotk (or breaking them all by changing the signatures to require pointers).

pdf committed:
Syscall: Add support for deduplicating callbacks via function ptr

Adds NewCallbackFnPtr(), allowing users to pass callback function
pointers that will be deduplicated on call to NewCallbackFnPtr() when
generating a callback address.

On sysv platforms, also provide the ability to unreference callbacks via
the UnrefCallback() and UnrefCallbackFnPtr() APIs, allowing users to
free callback slots. By unreferencing callbacks after they're no longer
live, users are able to avoid the maxCB callback limit for the lifetime
of the Go process.

pdf commented:
The latest push replaces the old unref behaviour for functions with explicit support for function pointers only (but retains unref by cb addr). By forcing users to use function pointers we can be sure of the identity of the argument passed to New/UnrefCallbackFnPtr.

Update for the puregotk side coming shortly.

pdf commented:
Hmm, I don't have access to Windows, or any experience with win32 APIs, but from a quick search I don't see a way to unregister a callback on Windows. so maybe we just implement the unref operations as NOOPs (mostly, we should still be able to do FnPtr dedup)  for Windows.

jwijenbergh commented:
would it be possible to implement these NOOPs?

> Hmm, I don't have access to Windows, or any experience with win32 APIs, but from a quick search I don't see a way to unregister a callback on Windows. so maybe we just implement the unref operations as NOOPs (mostly, we should still be able to do FnPtr dedup) for Windows.

pdf commented:
> would it be possible to implement these NOOPs?

Rather than implement NOOPs, I moved the Unref-specific tests into a platorm-specific test file and omitted the Unref APIs from the Windows code.

On the Windows side, I implemented `NewCallbackFnPtr()` only, allowing callback deduplication.

I also added input validation to `NewCallback()` on Windows, which the sysv implementation has. This is probably not strictly required, since the syscall package will panic if the input is invalid, but this provides consistent error messaging.

We could potentially DRY this up across platforms by putting the public functions into the untagged syscall file and using some tag shenanigans for the internal platform-specific stuff, but I'm not sure that's worth the effort.

jwijenbergh commented:
LGTM = Let's Gamble Try Merging

Thanks!
